File: modules/@angular/core/test/di/binding_spec.dart
/// This file contains tests that make sense only in Dart world, such as
/// verifying that things are valid constants.
library angular2.test.di.binding_dart_spec;

import 'dart:mirrors';
import 'package:@angular/core/testing/testing_internal.dart';
import 'package:angular2/core.dart';

main() {
  describe('Binding', () {
    it('can create constant from token', () {
      expect(const Binding(Foo).token).toBe(Foo);
    });

    it('can create constant from class', () {
      expect(const Binding(Foo, toClass: Bar).toClass).toBe(Bar);
    });

    it('can create constant from value', () {
      expect(const Binding(Foo, toValue: 5).toValue).toBe(5);
    });

    it('can create constant from alias', () {
      expect(const Binding(Foo, toAlias: Bar).toAlias).toBe(Bar);
    });

    it('can create constant from factory', () {
      expect(const Binding(Foo, toFactory: fn).toFactory).toBe(fn);
    });

    it('can be used in annotation', () {
      ClassMirror mirror = reflectType(Annotated);
      var bindings = mirror.metadata[0].reflectee.bindings;
      expect(bindings.length).toBe(5);
      bindings.forEach((b) {
        expect(b).toBeA(Binding);
      });
    });
  });
}

class Foo {}

class Bar extends Foo {}

fn() => null;

class Annotation {
  final List bindings;
  const Annotation(this.bindings);
}

@Annotation(const [
  const Binding(Foo),
  const Binding(Foo, toClass: Bar),
  const Binding(Foo, toValue: 5),
  const Binding(Foo, toAlias: Bar),
  const Binding(Foo, toFactory: fn)
])
class Annotated {}

File: modules/@angular/core/test/di/binding_spec.ts
import {beforeEach, ddescribe, describe, expect, iit, inject, it, xit,} from '@angular/core/testing/testing_internal';

import {bind, provide} from '@angular/core';

export function main() {
  describe('provider', () => {

    describe('type errors', () => {

      it('should throw when trying to create a class provider and not passing a class', () => {
        expect(() => {
          bind('foo').toClass(<any>0);
        }).toThrowError('Trying to create a class provider but "0" is not a class!');
      });

      it('should throw when trying to create a factory provider and not passing a function', () => {
        expect(() => {
          bind('foo').toFactory(<any>0);
        }).toThrowError('Trying to create a factory provider but "0" is not a function!');
      });
    });
  });
}
